### ISSUE TITLE
Document React Hot Loader 3

### ISSUE THREAD

**Issue by gaearon** [gaearon/react-hot-loader#243]
It features two modes (`react-hot-loader/babel` and `react-hot-loader/webpack`), doesn’t hide HMR API (you have to add `module.hot.accept` call yourself) and has a somewhat different API and setup process.

There’s info in https://github.com/gaearon/react-hot-boilerplate/pull/61 and https://github.com/gaearon/redux-devtools/commit/64f58b7010a1b2a71ad16716eb37ac1031f93915 but we should create good step-by-step docs and update the website too.

This will take a while because it’s not a priority for me. If you want to contribute, you are very welcome!
Obviously we can’t release 3.0 without any docs.

**Comment by gaearon** [comment]
To give some sense of perspective, if nobody contributes, the stable 3.0 might get out as late as June because I really don’t have the bandwidth to work on a good doc right now. It really is the biggest blocker to getting this version out!

**Comment by pablopaul** [comment]
@gaearon What kind of documentation is needed? Maybe you can give a rough documentation outline or so…

**Comment by gaearon** [comment]
We need to update [this intro](http://gaearon.github.io/react-hot-loader/getstarted/).

The new intro would be slightly longer because there is a bit more setup involved.
I would split it into several sections:

1. Enabling HMR
  * Create a development Webpack config separate from production one
  * Add `HotModuleReplacementPlugin` to development Webpack config
  * If you only render on the client, consider using `WebpackDevServer`
    * Easier to set up
    * Enable `hot: true` and add its entry points
  * If you use server rendering, consider using Express server + `webpack-dev-middleware`
    * More work but also more control
    * Show how to add `webpack-dev-middleware` and its entry point
2. Using HMR to replace the root component
  * Show how to add `module.hot` hook to re-render your root component
  * (This will not preserve state but is good enough for some cases!)
3. Adding React Hot Loader to preserve state
  * If you use Babel:
    * Add `react-hot-loader/babel` to `plugins`
  * If you don’t use Babel:
    * Add `react-hot-loader/webpack` to `loaders`
  * Add `react-hot-loader/patch` as the first entry point
  * Change your `index.js` to render `<AppContainer>` instead of your root component
  * That’s it!

**Comment by hedgerh** [comment]
I can help out with the docs.  I've been reading up and gathering as much info as I can.  I'm about to go through both the old and 3.0 source to try and figure out what's going on under the hood.  I'll see if I can get the ball rolling on the Intro, as well.

**Comment by gaearon** [comment]
Thanks!

**Comment by gadicc** [comment]
Can I suggest we get started on Hackpad first?  I think it will get us started quicker, and then we can use something like the [Hackpad downloader](https://chrome.google.com/webstore/detail/hackpad-downloader/jhkfiecgggifjcoliemlcajokapmcpih?hl=en) to download as markdown to commit, moving forward.

I took an initial bash at https://hackpad.com/React-Hot-Loader-v3-CxSDMncWUnb.

**Comment by gaearon** [comment]
@gadicc @hedgerh Can I ask you to work it out between yourselves? I added you both as collaborators so you should be able to work on a single PR together without much hassle. (Up to you if you want to first take it somewhere else, but I would prefer reviewing on GitHub.) Just a reminder that the PR(s) would need to be sent against the `next` branch. Thanks!

**Comment by tsaiDavid** [comment]
I'd be happy to help out with documentation!

**Comment by gadicc** [comment]
@gaearon, sure thing, thanks.

@hedgerh, sorry, I probably should have checked with you first on your progress but I just wanted to get the ball rolling.  But now we can collaborate together :)

*all*, we'll track this in https://github.com/gaearon/react-hot-loader/pull/257.

**Comment by gadicc** [comment]
@tsaiDavid, did you edit your post right like right now? :)  Sorry, I only saw it when refreshing the page, I had left the page open and your edit didn't come through.  Anyway, as you saw, everything is on GitHub now... your points are all great, would love you to submit it as a PR to `next-docs` as described in #257.

**Comment by gadicc** [comment]
@hedgerh, sorry, I also didn't notice your other PRs until right now!  I'll take a proper look tomorrow (it's night time here now), but maybe anything that's not in gh-pages we can merge into the `next-docs` branch and keep all in one place tracked in the single #257.

**Comment by hedgerh** [comment]
Hey hey @gadicc no worries.  I've been off all weekend so I'll need a little bit of time to get caught up.

My progress so far was:

- I prepped a migration guide in `docs/README.md`
- Removed all the starter kits bc I don't think they fit within the scope of the project
- I took a first pass on the intro that I'm not crazy about, but should be a decent start
- Added a little to the main readme

Totally open to any changes.  Just did as much research as I could and piece things together as I understood it.

**Comment by tsaiDavid** [comment]
@gadicc - sounds good I'll get right on it! @hedgerh happy to look at stuff for you guys, I love writing :)

**Comment by appsforartists** [comment]
In [next-docs](https://github.com/gaearon/react-hot-loader/tree/next-docs/docs#migration-to-30), Webpack is recommended as a fallback if you aren't already using Babel.  Is there a difference in featureset or reliability between the two adaptors, or is it entirely personal preference?

**Comment by gadicc** [comment]
@appsforartists, the webpack loader only registers/patches react components that are exported by the module/file, whereas the babel plugin works on all top level variable assignments.  This will be made clearer in the final version of the docs.

**Comment by appsforartists** [comment]
@gadicc That was my hunch.  Glad it will be documented. 😃

**Comment by hedgerh** [comment]
Hey @gaearon, just wanted to check in and see what the status on this was.  I haven't checked in on it in a couple of months.  I will take another look at them, but I think we just need you or someone that knows the lib to review the docs/site and see if anything needs to be added or removed.

Obviously no rush on my end.  Just wanted to touch base!

PR for docs: https://github.com/gaearon/react-hot-loader/pull/257
PR for github pages: https://github.com/gaearon/react-hot-loader/pull/260

**Comment by wmertens** [comment]
Question about the module accept code, the current examples are not DRY and risk differing setup between first render and hot reload. 

Why not take the function that does the hot accept and run it on first render too? Because of import graph analysis?

**Comment by janeklb** [comment]
@wmertens I agree, its a bit _wet_ so I've been doing something like:

```
import RootElement from './RootElement';

renderApp(RootElement);

// Hot Module Replacement API
if (module.hot) {
  module.hot.accept('./RootElement', () => {
    const RootElement = require('./RootElement').default;
    renderApp(RootElement);
  });
}

function renderApp(RootElement) {
  render(
    <AppContainer>
      <Provider store={store}>
        <RootElement />
      </Provider>
    </AppContainer>,
    document.getElementById('quickpick-root')
  );
}
```

**Comment by wmertens** [comment]
Nice, I like it! @gaearon something to make official?

BTW, WET stands for We Enjoy Typing 🙂

**Comment by janeklb** [comment]
@wmaurer never heard that one before -- excellent :)

**Comment by rvaidya** [comment]
A small optimization hack - I noticed that there are a bunch of places in my production javascript where my development env folder path is being included due to react-hot-loader.

If you use the Uglify plugin in your webpack production build you can add a global def in the Uglify config to drop this stuff out.

The react-hot-loader code itself could probably be modified to take better advantage of this by wrapping more of the module itself in if(__REACT_HOT_LOADER__) checks.

```
new webpack.optimize.UglifyJsPlugin({
	compress: {
		warnings: false,
		screw_ie8: true,
		drop_console: true,
		drop_debugger: true,
		dead_code: true,
		global_defs: {
			__REACT_HOT_LOADER__: undefined // eslint-disable-line no-undefined
		}
	}
}),
```

**Comment by calesce** [comment]
@rvaidya Is this the code from the Babel plugin? It should be omitted if you have `NODE_ENV=production`. See #367.

**Comment by rvaidya** [comment]
I'm on Windows so I was using DefinePlugin to define NODE_ENV.

Looks like I can achieve what you suggested by using a combination of DefinePlugin and cross-env.

**Comment by calesce** [comment]
@rvaidya Ah, forgot about Windows. :wink: Let me know if that doesn't work.

**Comment by rvaidya** [comment]
Using both cross env and DefinePlugin does work.

You'd think just using cross-env to set the environment variable would be enough, but my output file sizes are larger without the additional DefinePlugin entry setting process.env.

**Comment by calesce** [comment]
I believe `DefinePlugin` is needed for your own client-side checks, but env variables or `cross-env` for Babel plugins, other loaders and `node_modules`. I might be wrong on one of those, but yeah we should add those to our docs.

**Comment by patsissons** [comment]
I ran into a really challenging issue to diagnose today. I was importing my API surface first (which imports a bunch of stuff in the correct order), then importing the root component explicitly. This ended up creating a dependency graph that HMR could (or would) not handle. In order to resolve the issue, I had to import my API surface as a namespace and access the root component via the import alias.

This doesn't work:
```ts
declare let module: any;

import * as React from 'react';
import { render } from 'react-dom';
import { AppContainer } from 'react-hot-loader';

// import framework API surface
import './webrx-react';

// import Root Component
import { AppView } from './Components'; 

const container = document.getElementById('app');
render(<AppContainer><AppView /></AppContainer>, container);

if ((module as any).hot) {
  (module as any).hot.accept('./Components', () => {
    // ...
  }
}
```

but this does:
```ts
declare let module: any;

import * as React from 'react';
import { render } from 'react-dom';
import { AppContainer } from 'react-hot-loader';

// import framework API surface
import * as wxr from './webrx-react';

const container = document.getElementById('app');
render(<AppContainer><wxr.Components.AppView /></AppContainer>, container);

if ((module as any).hot) {
  (module as any).hot.accept('./webrx-react', () => {
    // ...
  }
}
```

I tried a number of combinations, but ultimately, if your root component has two parents it gets removed from `outdatedDependencies` and then we end up walking up the chain to module `0` which returns without a result.

I figured it's possible I'm not seeing something, so checking in with you guys.

**Comment by tleunen** [comment]
So I followed the steps to convert a project to the new "3.x" version and the page is realoading entirely every time a change is made. It wasn't the case before. Any idea on what could be the cause to that?

**Comment by calesce** [comment]
@tleunen want to open a new issue with a minimal reproduce project? There's a number of possibilities, with the most common being not pointing to RR routes in `module.hot.accept`.

I'm working on updating the troubleshooting doc with common 3.0-specific problems.

**Comment by tleunen** [comment]
I'm in a fairly big project so I can try to reproduce it in a smaller one but I'm setting the AppContainer  with the root component (which contains the redux provider and router).
```
if (module.hot) {
    module.hot.accept('./root', renderApp);
}
```

**Comment by alex-shamshurin** [comment]
I did as @gaearon advised, but I whatever I change I get `process-update.js:81[HMR] The following modules couldn't be hot updated: (Full reload needed) because the modules which have changed (and their parents) do not know how to hot reload themselves.`

My config
```
 entry     : {
    app: ['react-hot-loader/patch', hotMiddlewareScript, './bootstrap.js', 'babel-polyfill', './client']
  },
...
 plugins   : [
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NoErrorsPlugin(),
    new webpack.ProvidePlugin(configWebpack.providePlugin),
    new webpack.NormalModuleReplacementPlugin(/es6-promise$/, 'bluebird'),
    new webpack.DefinePlugin(configWebpack.definePlugin),
    new InlineEnviromentVariablesPlugin('NODE_ENV'),
  ],
```
babel: 

```
test   : /\.js$|\.jsx$/,
      loader : 'babel-loader',
      query  : function () {
        var presets = ['es2015', 'react', 'stage-0'];
        if (env.APP_ENV !== "development") {
          presets.push("react-optimize");
        }
        var plugins = [
          "transform-promise-to-bluebird",
          'transform-decorators-legacy',
          "transform-object-rest-spread",
          "transform-react-jsx-img-import"
        ]
        if (env.APP_ENV === "development" && !isServerConfig) {
          plugins.unshift("react-hot-loader/babel");
        }
        if (env.APP_ENV === "development") {
          plugins.push('transform-react-jsx-source');
          plugins.push('transform-react-jsx-self');
        }
        var cacheDirectory = env.APP_ENV === "development" && !isServerConfig;
        return {presets: presets, plugins: plugins, cacheDirectory}
      }(),
      include: path.join(__dirname, '..', 'app'),
      exclude: path.join(__dirname, '..', 'node_modules')
```

App.js

```
import {Provider} from 'react-redux';
import {Router} from 'react-router';
import WithStylesContext from './WithStylesContext';

function App(store, routes, history, onUpdate) {
  return (
    <Provider store={store}>
      <WithStylesContext onInsertCss={styles => styles._insertCss()}>
        <Router history={history} onUpdate={onUpdate}>
          {routes}
        </Router>
      </WithStylesContext>
    </Provider>
  );
}

export default App;
``
Client.js

```
...
const renderClient = (rootContainer, store, routes, history, onUpdate) => {
  render(
    <AppContainer>
      {rootContainer(store, routes, history, onUpdate)}
    </AppContainer>, document.getElementById('app'), () => {
      if (initialRendering) {
        const node = document.getElementById('server-css');
        if (node) {
          node.parentNode.removeChild(node);
          initialRendering = false;
        }
      }
    }
  );
}

if (module.hot) {
  module.hot.accept('./App.jsx', () => {
    const NextRootContainer = require('./App.jsx').default;
    renderClient(NextRootContainer, store, routes, history, onUpdate);
  })
}

renderClient(App, store, routes, history, onUpdate);
```

What else did I miss?

**Comment by alex-shamshurin** [comment]
I dublicate this https://github.com/gaearon/react-hot-loader/issues/433

**Comment by wkwiatek** [comment]
We just merged #260 so I think we can close this general issue for now. If you can see that something in the docs is not right or there is some room for improvements, feel free to open either issue or PR. 

Thanks!
